### PR TITLE
Cryptoconditions (PREIMAGE-SHA-256 & ED25519)

### PR DESCRIPTION
**Pull request by nbougalis**
Cryptoconditions provide a mechanism to describe a signed message such that multiple actors in a distributed system can all verify the same signed message and agree on whether it matches the description. This provides a useful primitive for event-based systems that are distributed on the Internet since we can describe events in a standard deterministic manner (represented by signed messages) and therefore define generic authenticated event handlers.

### CODE REVIEW

**Comment by seelabs** [comment]
@nbougalis You can cherry pick this to fix the non-unity build. It also fixes a warning from clang (it's a bogus warning, but I want to silence it anyway): https://github.com/seelabs/rippled/commit/c86122eaaca0e41155f5709b487a844199b4a927

File: src/ripple/conditions/condition.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+// NIKB-TODO: These should be `enum class : std::uint32_t`
+std::uint32_t constexpr feature_sha256             = 1;
+std::uint32_t constexpr feature_preimage           = 2;
+std::uint32_t constexpr feature_prefix             = 4;
+std::uint32_t constexpr feature_threshold          = 8;
+std::uint32_t constexpr feature_rsa_pss            = 16;
+std::uint32_t constexpr feature_ed25519            = 32;
+
+struct condition_t
```
Comment (seelabs) [inline src/ripple/conditions/condition.h]: Type names and variable names don't follow ripple coding standards. I understand this was going to be a stand-alone lib, hence the style conflict.

File: src/ripple/conditions/condition.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+{
+    std::uint16_t type;
+    std::uint16_t fulfillment_length;
```
Comment (seelabs) [inline src/ripple/conditions/condition.h]: This is `max_fulfillment_length`. Rename?

File: src/ripple/conditions/condition.h [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+{
+    if (c.fulfillment_length > 65535)
+        throw std::length_error ("The maximum fulfillment length supported is 65535");
+
+    if (c.type != condition_hashlock &&
+            c.type != condition_prefix_sha256 &&
+                c.type != condition_threshold_sha256 &&
+                    c.type != condition_rsa_sha256 &&
+                        c.type != condition_ed25519)
+        Throw<std::logic_error>("Unknown fulfillment type");
```
Comment (seelabs) [inline src/ripple/conditions/condition.h]: What do you think about:
```
    switch (c.type)
    {
        case condition_hashlock:
        case condition_prefix_sha256:
        case condition_threshold_sha256:
        case condition_rsa_sha256:
        case condition_ed25519:
            break;
        default:
            Throw<std::logic_error>("Unknown fulfillment type");
    }
```

File: src/ripple/conditions/ed25519.h [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    static std::size_t constexpr pubkey_size_ = 32;
+
+    std::array<std::uint8_t,
+        pubkey_size_ + signature_size_> payload_;
+
+public:
+    ed25519_t () = default;
+
+    /** Create a fulfillment */
+    ed25519_t (std::vector<std::uint8_t> const& payload)
```
Comment (seelabs) [inline src/ripple/conditions/ed25519.h]: `explicit`

File: src/ripple/conditions/fulfillment.h [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+bool
+operator!= (fulfillment_t const& lhs, fulfillment_t const& rhs)
+{
+    return !(lhs == rhs);
+}
+
+/** Load a serialized condition either from its string or binary form */
+/** @{ */
+std::unique_ptr<fulfillment_t>
+load_fulfillment (std::string s);
```
Comment (seelabs) [inline src/ripple/conditions/fulfillment.h]: `std::string const&`

File: src/ripple/conditions/hashlock.h [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+#include <boost/regex.hpp>
+#include <boost/optional.hpp>
+#include <algorithm>
+#include <iterator>
+#include <vector>
+
+namespace ripple {
+namespace cryptoconditions {
+
+struct hashlock_t final
```
Comment (seelabs) [inline src/ripple/conditions/hashlock.h]: We tend to use `final` very sparingly. Unless this _needs_ to be `final` it probably fits in better with rippled if we remove the keyword.

File: src/ripple/conditions/hashlock.h [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    : public fulfillment_t
+{
+    Buffer payload_;
+
+public:
+    hashlock_t (std::vector<std::uint8_t> payload)
```
Comment (seelabs) [inline src/ripple/conditions/hashlock.h]: `std::vector<std::uint8_t> const&`

File: src/ripple/conditions/impl/condition.cpp [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    // more work and probability of error.
+
+    // TODO: use two regex: one that accepts anything the
+    // standard supports and one which accepts only what
+    // we support. Parse with both for improved errors?
+    static boost::regex const re_current (
+            "^"                            // start of line
+            "cc\\:"                        // 'cc' for cryptocondition
+            "([1-9a-f][0-9a-f]{0,3}|0):"   // type (hexadecimal)
+            "([1-9a-f][0-9a-f]{0,15}):"    // feature bitmask (hexadecimal)
```
Comment (seelabs) [inline src/ripple/conditions/impl/condition.cpp]: We specify 16 hex digits for the feature bitmask, but we're putting it into a 32 bit number with no overflow check (line 62)

File: src/ripple/conditions/impl/condition.cpp [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    if (!boost::regex_match (s, match, re_current))
+        return boost::none;
+
+    try
+    {
+        condition_t c;
+
+        c.type = parse_hexadecimal<std::uint16_t> (match[1]);
+        c.requires = parse_hexadecimal<std::uint32_t>(match[2]);
+        c.fulfillment_length = parse_decimal<std::uint16_t>(match[4]);
```
Comment (seelabs) [inline src/ripple/conditions/impl/condition.cpp]: Fulfillment length can be lg(10^18) ~= 60 bits log, and we're putting it into a 16 bit number with no overflow check.

File: src/ripple/conditions/impl/condition.cpp [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+boost::optional<condition_t>
+load_condition(std::string const& s)
+{
+    // CHECKME: custom parser maybe? probably faster but
+    // more work and probability of error.
+
+    // TODO: use two regex: one that accepts anything the
+    // standard supports and one which accepts only what
+    // we support. Parse with both for improved errors?
+    static boost::regex const re_current (
```
Comment (seelabs) [inline src/ripple/conditions/impl/condition.cpp]: I _think_ all our platforms support `std::regex`. I'd prefer to use things from `std` before `boost`.

File: src/ripple/conditions/impl/condition.cpp [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            "^"                            // start of line
+            "cc\\:"                        // 'cc' for cryptocondition
+            "([1-9a-f][0-9a-f]{0,3}|0):"   // type (hexadecimal)
```
Comment (seelabs) [inline src/ripple/conditions/impl/condition.cpp]: Hexadecimal digits can't be capitalized? i.e. `FF` is invalid?

File: src/ripple/conditions/impl/condition.cpp (hunk 8 above)
Comment (seelabs) [inline src/ripple/conditions/impl/condition.cpp]: You might be able to use the `[:xdigit:]` character class here (would have to check about case sensitivity and which `std::regex` grammars supports it.

File: src/ripple/conditions/impl/condition.cpp [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    }
+    catch (std::exception const& x)
+    {
+        return boost::none;
+    }
+}
+
+std::string
+to_string (condition_t const& c)
+{
```
Comment (seelabs) [inline src/ripple/conditions/impl/condition.cpp]: Is it worth optimizing this to reduce memory allocations? We could reserve in a result string up front and then `operator+=` or `append` into it.

File: src/ripple/conditions/impl/fulfillment.cpp [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+validate (
+    fulfillment_t const& f,
+    Slice const& m)
+{
+    return validate (f, f.condition(), m);
+}
+
+// Parse a condition from its string form
+std::unique_ptr<fulfillment_t>
+load_fulfillment (std::string s)
```
Comment (seelabs) [inline src/ripple/conditions/impl/fulfillment.cpp]: `std::string const&`

File: src/ripple/conditions/impl/fulfillment.cpp [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    // more work and probability of error.
+
+    // TODO: use two regex: one that accepts anything the
+    // standard supports and one which accepts only what
+    // we support. Parse with both for improved errors?
+    static boost::regex const re_current (
+            "^"                            // start of line
+            "cf\\:"                        // 'cf' for fulfillment
+            "([1-9a-f][0-9a-f]{0,3}|0):"   // type
+            "([a-zA-Z0-9_-]*)"             // fulfillment payload (base64url)
```
Comment (seelabs) [inline src/ripple/conditions/impl/fulfillment.cpp]: Payload can be empty?

File: src/ripple/conditions/impl/fulfillment.cpp [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+std::unique_ptr<fulfillment_t>
+load_fulfillment (std::string s)
+{
+    // CHECKME: custom parser maybe? probably faster but
+    // more work and probability of error.
+
+    // TODO: use two regex: one that accepts anything the
+    // standard supports and one which accepts only what
+    // we support. Parse with both for improved errors?
+    static boost::regex const re_current (
```
Comment (seelabs) [inline src/ripple/conditions/impl/fulfillment.cpp]: Same comment as the other regex: Use `std::regex`, maybe use `[:xdigit:]`.

File: src/ripple/conditions/condition.h (hunk 2 above)
Comment (nbougalis) [inline src/ripple/conditions/condition.h]: I don't follow?

File: src/ripple/conditions/impl/condition.cpp (hunk 12 above)
Comment (nbougalis) [inline src/ripple/conditions/impl/condition.cpp]: Yes: this will be optimized.

File: src/ripple/conditions/impl/fulfillment.cpp (hunk 14 above)
Comment (nbougalis) [inline src/ripple/conditions/impl/fulfillment.cpp]: Yes, see _inter alia_, https://github.com/nbougalis/rippled/blob/64d3d08ec2886aed4aeb2a0bc8f2f85dfdb32759/src/ripple/conditions/tests/hashlock_test.cpp#L71-L73

**Comment by nbougalis** [comment]
I'm going to close this to address the issues @seelabs identifies and finish the binary OER encoding, then reopen. Thanks.

File: src/ripple/conditions/condition.h (hunk 2 above)
Comment (seelabs) [inline src/ripple/conditions/condition.h]: I thought the field was the maximum length of a fullfillment. I.e. "reject any fullfillment greater than this", not "reject any fulfillment not equal to this". Did I misunderstand the point of this member?

**Pull request closed without merging**